horovod/torch/adapter.cc

```diff
@@ -0,0 +1,158 @@
+#endif
+
+#include "adapter.h"
+#include "cuda_util.h"
+#include "tensor_util.h"
+
+namespace horovod {
+namespace torch {
+
+// This class intentionally does not have destructor at the moment.
+//
+// Unfortunately, by the time this destructor would be called in normal
+// circumstances (application shutdown), CUDA context would already be destroyed
+// and cudaFree() operations would print nasty errors in the log - in a pretty
+// normal termination scenario.
+//
+// If we add functionality to terminate Horovod without terminating the
+// application, we should revisit this logic.
+TorchPersistentBuffer::TorchPersistentBuffer(int device, int64_t size)
+    : device_(device) {
+  with_device device_context(device_);
+  if (device_ == CPU_DEVICE_ID) {
+    buffer_ = new char[size];
+  } else {
+#if HAVE_CUDA
+    THCudaCheck(THCudaMalloc(state, (void**)&buffer_, size));
+#else
+    throw std::logic_error("Internal error. Requested TorchPersistentBuffer "
+                           "with GPU device but not compiled with CUDA.");
+#endif
+  }
+}
+
+const void*
+TorchPersistentBuffer::AccessData(std::shared_ptr<OpContext> context) const {
+  return buffer_;
+}
+
+template <class T> TorchTensor<T>::TorchTensor(T* tensor) : tensor_(tensor) {}
+
+template <class T> const MPIDataType TorchTensor<T>::dtype() const {
+  return TensorUtil::GetDType<T>();
+}
+
+template <class T> const TensorShape TorchTensor<T>::shape() const {
+  auto shape = TensorUtil::GetShape(tensor_);
+  if (shape.dims() == 0) {
+    // Tensor with empty shape is a Tensor with no values in PyTorch, unlike a
+    // constant in TensorFlow. So, we inject a dummy zero dimension to make sure
+    // that the number-of-elements calculation is correct.
+    shape.AddDim(0);
+  }
+  return shape;
+}
+
+template <class T> const void* TorchTensor<T>::data() const {
+  return TensorUtil::GetData(tensor_);
+}
+
+template <class T> int64_t TorchTensor<T>::size() const {
+  return TensorUtil::GetSize(tensor_);
+}
+
+template <class T>
+TorchTemporaryBuffer<T>::TorchTemporaryBuffer(int device)
+    : TorchTensor<T>(nullptr) {
+  this->tensor_ = TensorUtil::New<T>(device);
+}
+
+template <class T> TorchTemporaryBuffer<T>::~TorchTemporaryBuffer() {
+  TensorUtil::Free(this->tensor_);
+}
+
+template <class T> T* TorchTemporaryBuffer<T>::tensor() const {
+  return this->tensor_;
+}
+
+template <class T>
+TorchOpContext<T>::TorchOpContext(int device, T* output)
+    : device_(device), output_(output) {}
+
+template <class T>
+Status TorchOpContext<T>::AllocatePersistent(
+    int64_t size, std::shared_ptr<PersistentBuffer>* tensor) {
+  // Allocation errors are handled using PyTorch exceptions.
+  *tensor = std::make_shared<TorchPersistentBuffer>(device_, size);
+  return Status::OK();
+}
+
+template <class T>
+Status TorchOpContext<T>::AllocateOutput(TensorShape shape,
+                                         std::shared_ptr<Tensor>* tensor) {
+  int64_t* shape_array = new int64_t[shape.dims()];
```

Comment:
use vector

Reply:
Underlying function `THTensor##_resizeNd(tensor, nDimension, size, stride);` uses an array pointer, so I'm just saving conversion back and forth.